File: app/api/db/models/Community.py
from django.db import models

from typing import List

from .Image import ImageModel

from .Page import PageModel

from .Event import EventModel

from .Department import DepartmentModel

from .TranslatableContent import TranslatableContentModel

from ....entities import Community

class CommunityModel(models.Model):

	tag             = models.CharField(max_length = 8, unique = True, primary_key=True)

	name            = models.CharField(max_length = 100, unique = True)

	description_ids = models.ManyToManyField(TranslatableContentModel, blank = True)

	images_ids      = models.ManyToManyField(ImageModel, blank = True)

	pages_types_ids = models.ManyToManyField(PageModel, blank = True)

	events_ids      = models.ManyToManyField(EventModel, blank = True)

	departments_ids = models.ManyToManyField(DepartmentModel, blank = True)

	def description(self, lang : str) -> List[str]:

		description = [content_model.content(lang) for content_model in self.description_ids.all()]

		return description

	def images(self) -> List[ImageModel]:

		return self.images_ids.all()

	def events(self) -> List[EventModel]:

		return self.events_ids.all()

	def departments(self) ->List[DepartmentModel]:

		return self.departments_ids.all()

	def pages(self) -> List[PageModel]:

		return self.pages_types_ids.all()

	def toEntity(self, lang : str) -> Community:

		images = [image.url() for image in self.images()]

		pages = [page.toEntity(lang) for page in self.pages()]

		return Community(self.name, self.tag, self.description(lang), images, pages)

	def __str__(self):

		return self.name

File: static/src/api/storage/modules/community.ts

// @ts-ignore
import { setStore, getStore, } from '../handler.ts';

// @ts-ignore
import { Community, Event, Page, } from '../../entities.ts';

export default class CommunityModule {

	private _community: Community | null = null;

	public constructor(tag: string) {
		this._community = getStore(tag);
	}

	public static addCommunity(community: Community) { setStore(community.tag, JSON.stringify(community), 500); }

	public get community() : Community { return this._community; }

	public get name()   : string         { return this._community.name; }
	public get description() : string[]  { return this._community.description; }
	public get pages()  : Page[]         { return this._community.pages; }
	public get tag()    : string         { return this._community.tag; }
	public get events() : Event[]        { return this._community.events; }

	public setCommunity(community: Community) { this._community = community; setStore(community.tag, JSON.stringify(this._community), 500); }

	public addEvents(events: Event[]) { this._community.events = events; setStore(this._community.tag, JSON.stringify(this._community), 500); }
}
